### ISSUE TITLE
Accuracy issues for outputs with double precision

### ISSUE THREAD

**Issue by qedawkins** [iree-org/iree#11924]
### What happened?

The following IR gives incorrect output (first showing through vulkan)
```mlir
#map = affine_map<(d0, d1, d2, d3) -> (0, d1, d2, d3)>
#map1 = affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>
#map2 = affine_map<(d0, d1, d2, d3) -> ()> 
builtin.module {
  func.func @forward(%arg0: tensor<1x2x32x32xf64>) -> tensor<1x2x32x32xf64> {
    %cst_0 = arith.constant dense<5.0e-01> : tensor<f64>
    %2 = tensor.empty() : tensor<1x2x32x32xf64>
    %4 = linalg.generic {indexing_maps = [#map, #map2, #map1], iterator_types = ["parallel", "parallel", "parallel", "parallel"]} ins(%arg0, %cst_0 : tensor<1x2x32x32xf64>, tensor<f64>) outs(%2 : tensor<1x2x32x32xf64>) {
    ^bb0(%in: f64, %in_2: f64, %out: f64):
      %5 = arith.mulf %in, %in_2 : f64 
      linalg.yield %5 : f64 
    } -> tensor<1x2x32x32xf64>
    return %4 : tensor<1x2x32x32xf64>
  }
}

```
```
iree-compile --iree-input-type=none --iree-hal-target-backends=vulkan --iree-vulkan-target-triple=rdna3-unknown-linux --iree-stream-resource-index-bits=64 --iree-vm-target-index-bits=64 int64.mlir -o int64.vmfb
iree-run-module --module_file=int64.vmfb --entry_function=forward --device=vulkan --function_input=1x2x32x32xf64=1.0 --print_max_element_count=8192
```
```
result[0]: hal.buffer_view
1x2x32x32xf64=[[[0.00390625 0.00390625 0.00390625 ...]][[1 1 1 1 ...]]]
```
where the correct per-element value is 0.5. Note that to get the values to print I had to apply the following hack
```
diff --git a/runtime/src/iree/hal/buffer.c b/runtime/src/iree/hal/buffer.c
index 4ecffc376..56b152481 100644
--- a/runtime/src/iree/hal/buffer.c
+++ b/runtime/src/iree/hal/buffer.c
@@ -419,7 +419,7 @@ static iree_status_t iree_hal_buffer_calculate_range(
   }
 
   *out_adjusted_offset = adjusted_offset;
-  if (out_adjusted_length) *out_adjusted_length = adjusted_length;
+  if (out_adjusted_length) *out_adjusted_length = adjusted_length * 2;
   return iree_ok_status();
 }
```
 to get around an OUT_OF_RANGE error
```
EXEC @forward
OUT_OF_RANGE; printing results
```

When trying to run through cpu however, I hit the following (which may be from the hack so I'm unsure whether this is a real error) but either way I was unable to get values to print with iree-run-module.
```
malloc(): corrupted top size
./int64_fail.sh: line 5: 685488 Aborted                 (core dumped) ~/nod/iree-build/tools/iree-run-module --module_file=samples/int64.vmfb --entry_function=forward --device=vulkan --function_input=1x2x32x32xf64=1.0 --print_max_element_count=8192
```

This IR came from a slightly larger example starting from torch-mlir where we can see a similar result for CPU
```python
import torch_mlir
from torch_mlir_e2e_test.linalg_on_tensors_backends import refbackend
import torch
from brevitas.nn import QuantIdentity # pip install git+https://github.com/Xilinx/brevitas.git@dev
from brevitas.export import export_torch_qcdq 
from brevitas.inject.enum import *


class Model(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.identity = QuantIdentity(return_quant_tensor=True)
    
    def forward(self, x): 
        x = self.identity(x)[0]
        return x

model = Model()
model.eval()

example_input = torch.ones(1, 2, 32, 32) 

print("PYTORCH OUTPUT: ")
print(model(example_input))
traced_model = export_torch_qcdq(model, args=torch.randn(1, 2, 32, 32), export_path='test_brevitas.pt')

linalg_on_tensors_mlir = torch_mlir.compile(
    traced_model, [torch.randn(1, 2, 32, 32),], output_type=torch_mlir.OutputType.LINALG_ON_TENSORS,
    use_tracing=False, ignore_traced_shapes=False)
print(linalg_on_tensors_mlir)

import iree.runtime as ireert
from iree.runtime import get_driver, get_device
import iree.compiler as ireec

device = "llvm-cpu"
runtime_device = "local-sync"

flatbuffer = ireec.compile_str(
    str(linalg_on_tensors_mlir),
    target_backends=[device],
)

config = ireert.Config(driver_name=runtime_device)
vm_module = ireert.VmModule.from_flatbuffer(config.vm_instance, flatbuffer)
ctx = ireert.SystemContext(config=config)
ctx.add_vm_module(vm_module)
forward = ctx.modules.module["forward"]

result = forward(example_input.numpy()).to_host()
print("IREE OUTPUT: ")
print(result)

backend = refbackend.RefBackendLinalgOnTensorsBackend()
compiled = backend.compile(linalg_on_tensors_mlir)
jit_module = backend.load(compiled)
print("REFBACKEND OUTPUT: ")
print(jit_module.forward(example_input.numpy()))
```
Where the linalg produced by this example looks like the following
```mlir
#map = affine_map<(d0, d1, d2, d3) -> (0, d1, d2, d3)>
#map1 = affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>
#map2 = affine_map<(d0, d1, d2, d3) -> ()>
module attributes {torch.debug_module_name = "_JitTraceExportWrapper"} {
  ml_program.global private mutable @global_seed(dense<0> : tensor<i64>) : tensor<i64>
  func.func @forward(%arg0: tensor<1x2x32x32xf32>) -> tensor<1x2x32x32xf64> {
    %cst = arith.constant dense<0.000000e+00> : tensor<f64>
    %cst_0 = arith.constant dense<7.812500e-03> : tensor<f64>
    %c-128_i32 = arith.constant -128 : i32
    %c127_i32 = arith.constant 127 : i32
    %cst_1 = arith.constant 7.812500e-03 : f32
    %0 = tensor.empty() : tensor<1x2x32x32xi8>
    %1 = linalg.generic {indexing_maps = [#map, #map1], iterator_types = ["parallel", "parallel", "parallel", "parallel"]} ins(%arg0 : tensor<1x2x32x32xf32>) outs(%0 : tensor<1x2x32x32xi8>) {
    ^bb0(%in: f32, %out: i8):
      %5 = arith.divf %in, %cst_1 : f32
      %6 = math.round %5 : f32
      %7 = arith.fptosi %6 : f32 to i32
      %8 = arith.cmpi slt, %7, %c-128_i32 : i32
      %9 = arith.select %8, %c-128_i32, %7 : i32
      %10 = arith.cmpi sgt, %9, %c127_i32 : i32
      %11 = arith.select %10, %c127_i32, %9 : i32
      %12 = arith.trunci %11 : i32 to i8
      linalg.yield %12 : i8
    } -> tensor<1x2x32x32xi8>
    %2 = tensor.empty() : tensor<1x2x32x32xf64>
    %3 = linalg.generic {indexing_maps = [#map, #map2, #map1], iterator_types = ["parallel", "parallel", "parallel", "parallel"]} ins(%1, %cst : tensor<1x2x32x32xi8>, tensor<f64>) outs(%2 : tensor<1x2x32x32xf64>) {
    ^bb0(%in: i8, %in_2: f64, %out: f64):
      %5 = arith.sitofp %in : i8 to f64
      %6 = arith.subf %5, %in_2 : f64
      linalg.yield %6 : f64
    } -> tensor<1x2x32x32xf64>
    %4 = linalg.generic {indexing_maps = [#map, #map2, #map1], iterator_types = ["parallel", "parallel", "parallel", "parallel"]} ins(%3, %cst_0 : tensor<1x2x32x32xf64>, tensor<f64>) outs(%2 : tensor<1x2x32x32xf64>) {
    ^bb0(%in: f64, %in_2: f64, %out: f64):
      %5 = arith.mulf %in, %in_2 : f64
      linalg.yield %5 : f64
    } -> tensor<1x2x32x32xf64>
    return %4 : tensor<1x2x32x32xf64>
  }
}
```
and we see the values from the torch-mlir reference backend are correct
```
PYTORCH OUTPUT: 
tensor([[[[0.9922, 0.9922, 0.9922,  ..., 0.9922, 0.9922, 0.9922],
          [0.9922, 0.9922, 0.9922,  ..., 0.9922, 0.9922, 0.9922],
          [0.9922, 0.9922, 0.9922,  ..., 0.9922, 0.9922, 0.9922],
          ...,
          [0.9922, 0.9922, 0.9922,  ..., 0.9922, 0.9922, 0.9922],
          [0.9922, 0.9922, 0.9922,  ..., 0.9922, 0.9922, 0.9922],
          [0.9922, 0.9922, 0.9922,  ..., 0.9922, 0.9922, 0.9922]],

         [[0.9922, 0.9922, 0.9922,  ..., 0.9922, 0.9922, 0.9922],
          [0.9922, 0.9922, 0.9922,  ..., 0.9922, 0.9922, 0.9922],
          [0.9922, 0.9922, 0.9922,  ..., 0.9922, 0.9922, 0.9922],
          ...,
          [0.9922, 0.9922, 0.9922,  ..., 0.9922, 0.9922, 0.9922],
          [0.9922, 0.9922, 0.9922,  ..., 0.9922, 0.9922, 0.9922],
          [0.9922, 0.9922, 0.9922,  ..., 0.9922, 0.9922, 0.9922]]]])
IREE OUTPUT: 
[[[[ 7.32421967e-003  7.32421967e-003  7.32421967e-003 ...
     7.32421967e-003  7.32421967e-003  7.32421967e-003]
   [ 7.32421967e-003  7.32421967e-003  7.32421967e-003 ...
     7.32421967e-003  7.32421967e-003  7.32421967e-003]
   [ 7.32421967e-003  7.32421967e-003  7.32421967e-003 ...
     7.32421967e-003  7.32421967e-003  7.32421967e-003]
   ...
   [ 7.32421967e-003  7.32421967e-003  7.32421967e-003 ...
     7.32421967e-003  7.32421967e-003  7.32421967e-003]
   [ 7.32421967e-003  7.32421967e-003  7.32421967e-003 ...
     7.32421967e-003  7.32421967e-003  7.32421967e-003]
   [ 7.32421967e-003  7.32421967e-003  7.32421967e-003 ...
     7.32421967e-003  7.32421967e-003  7.32421967e-003]]

  [[ 0.00000000e+000  0.00000000e+000  0.00000000e+000 ...
     1.48219694e-323  4.24399166e-314  0.00000000e+000]
   [ 1.48219694e-323  8.48798321e-314  0.00000000e+000 ...
     0.00000000e+000  1.48219694e-323  1.48539706e-313]
   [ 0.00000000e+000  1.48219694e-323  4.24399166e-314 ...
     1.90979622e-313  0.00000000e+000  1.48219694e-323]
   ...
   [ 6.91873820e-310  0.00000000e+000  0.00000000e+000 ...
     0.00000000e+000 -7.47679484e+091  1.08721759e-314]
   [ 0.00000000e+000  0.00000000e+000  1.90766232e-314 ...
    -3.92643185e-281  4.65963949e-310  4.65963936e-310]
   [ 2.47032823e-323  0.00000000e+000  0.00000000e+000 ...
                 nan  9.00103463e-315  4.65963951e-310]]]]

REFBACKEND OUTPUT: 
[[[[0.9921875 0.9921875 0.9921875 ... 0.9921875 0.9921875 0.9921875]
   [0.9921875 0.9921875 0.9921875 ... 0.9921875 0.9921875 0.9921875]
   [0.9921875 0.9921875 0.9921875 ... 0.9921875 0.9921875 0.9921875]
   ...
   [0.9921875 0.9921875 0.9921875 ... 0.9921875 0.9921875 0.9921875]
   [0.9921875 0.9921875 0.9921875 ... 0.9921875 0.9921875 0.9921875]
   [0.9921875 0.9921875 0.9921875 ... 0.9921875 0.9921875 0.9921875]]

  [[0.9921875 0.9921875 0.9921875 ... 0.9921875 0.9921875 0.9921875]
   [0.9921875 0.9921875 0.9921875 ... 0.9921875 0.9921875 0.9921875]
   [0.9921875 0.9921875 0.9921875 ... 0.9921875 0.9921875 0.9921875]
   ...
   [0.9921875 0.9921875 0.9921875 ... 0.9921875 0.9921875 0.9921875]
   [0.9921875 0.9921875 0.9921875 ... 0.9921875 0.9921875 0.9921875]
   [0.9921875 0.9921875 0.9921875 ... 0.9921875 0.9921875 0.9921875]]]]

```

**Comment by MaheshRavishankar** [comment]
I am not sure this is an IREE compilation issue per se.... there is a lot of bitwidth changes and int to float conversions happening. There is a pass in IREE that demotes f64 to f32... that might be introducing some errors here....

**Comment by qedawkins** [comment]
Agreed, it's not clear to me that this is a compilation error so I didn't mark it as such. The pass to demote f64 to f32 doesn't catch the output though which I assume is leading to some weird results
```mlir
    %5 = linalg.generic {indexing_maps = [#map, #map2, #map1], iterator_types = ["parallel", "parallel", "parallel", "parallel"]} ins(%4, %cst_0 : tensor<1x2x32x32xf32>, tensor<f32>) outs(%3 : tensor<1x2x32x32xf32>) {
    ^bb0(%in: f32, %in_2: f32, %out: f32): 
      %7 = arith.mulf %in, %in_2 : f32 
      linalg.yield %7 : f32 
    } -> tensor<1x2x32x32xf32>
    %6 = hal.tensor.export %5 : tensor<1x2x32x32xf64> as tensor<1x2x32x32xf32> -> !hal.buffer_view
    return %6 : !hal.buffer_view
```
but using iree-opt to apply the pattern as a preprocessing step gives fully f32 code that gives correct results
```
iree-opt quant_fail.mlir --iree-util-demote-f64-to-f32 -o quant_fail_f32.mlir
```
```
EXEC @forward
result[0]: hal.buffer_view
1x2x32x32xf32=[[[0.992188 0.992188 0.992188 0.992188 0.992188 0.992188
```

**Comment by allieculp** [comment]
@qedawkins Is this still active?

**Comment by qedawkins** [comment]
This is probably still an issue (and might be linked with #12967) but to my knowledge it is not a part of any critical work and is just something I came across.

**Comment by allieculp** [comment]
Got it, thanks. I'm going to move this to the backlog for now, please bring back up if needed!

**Comment by benvanik** [comment]
Original input does not reproduce with llvm-cpu or vulkan with the `--iree-input-demote-f64-to-f32=false` flag. I suspect the issue (and why multiplying the size * 2 stopped the errors) was that things were getting demoted in unexpected ways.

If I don't pass the flag and try to pass an f64 tensor I get the right error:
```
D:\Dev\iree\runtime\src\iree\modules\hal\utils\buffer_diagnostics.c:191: INVALID_ARGUMENT; input0 element type mismatch; expected f32 (21000020) but have f64 (21000040); stack:
  0x00007ff6c986200a iree-run-module <iree_hal_modules_buffer_view_assert+0x2da> (D:\Dev\iree\runtime\src\iree\modules\hal\utils\buffer_diagnostics.c:185)
  0x00007ff6c9814173 iree-run-module <iree_hal_module_buffer_view_assert+0x1c3> (D:\Dev\iree\runtime\src\iree\modules\hal\module.c:613)
```

and if I pass an f32 as expected with demotion I also get the correct results.
